File: example/jchatbot/src/main/java/zw/co/dcl/jchatbot/service/WebhookConfigService.java
package zw.co.dcl.jchatbot.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import zw.co.dcl.jawce.engine.model.SessionSettings;
import zw.co.dcl.jawce.engine.model.WhatsappSettings;
import zw.co.dcl.jawce.engine.model.dto.ChannelOriginConfig;
import zw.co.dcl.jawce.engine.model.dto.EngineRequestSettings;
import zw.co.dcl.jawce.engine.model.dto.WaEngineConfig;
import zw.co.dcl.jawce.engine.service.EntryService;
import zw.co.dcl.jawce.session.ISessionManager;
import zw.co.dcl.jchatbot.configs.ChannelConfig;
import zw.co.dcl.jchatbot.configs.ChatbotConfig;
import zw.co.dcl.jchatbot.configs.TemplateConfig;

import java.util.ArrayList;

@Service
public class WebhookConfigService {

    private final ISessionManager sessionManager;
    private final ChatbotConfig config;
    private final TemplateConfig templateConfig;
    private final RestTemplate restTemplate;
    private final ChannelConfig externalConfig;

    @Value("${resources.hooks.base-url}")
    private String botEngineHookBaseUrl;
    @Value("${resources.hooks.security-token}")
    private String botEngineHookUrlToken;

    public WebhookConfigService(
            ISessionManager sessionManager,
            ChatbotConfig config,
            TemplateConfig templateConfig,
            RestTemplate restTemplate,
            ChannelConfig externalConfig
    ) {
        this.sessionManager = sessionManager;
        this.config = config;
        this.templateConfig = templateConfig;
        this.restTemplate = restTemplate;
        this.externalConfig = externalConfig;
    }

    public EntryService getEntryInstance() {
        var channelOrigin = new ChannelOriginConfig(
                false,
                new ArrayList<>(),
                false,
                null,
                "*"
        );

        return EntryService.getInstance(engineConfig(), channelOrigin);
    }

    private WaEngineConfig engineConfig() {
        return new WaEngineConfig(
                sessionManager,
                templateConfig.getBotTemplates(),
                templateConfig.getBotTriggers(),
                new EngineRequestSettings(
                        botEngineHookBaseUrl,
                        botEngineHookUrlToken
                ),
                whatsappChannelSettings(),
                sessionSettings(),
                restTemplate
        );
    }

    private WhatsappSettings whatsappChannelSettings() {
        var settings = new WhatsappSettings();

        settings.setHubToken(config.getHubToken());
        settings.setAccessToken(config.getAccessToken());
        settings.setPhoneNumberId(config.getPhoneNumberId());
        settings.setLocalTesting(externalConfig.getTestLocal().booleanValue());
        settings.setLocalTestingRequestUrl(externalConfig.getLocalUrl());
        settings.setApiVersion("v21.0");

        return settings;
    }

    private SessionSettings sessionSettings() {
        var settings = new SessionSettings();
        settings.setSessionTTL(config.getSessionTtl());
        settings.setHandleSessionInactivity(false);
        settings.setStartMenuStageKey(config.getInitialStage());
        return settings;
    }
}

File: chatbot/src/main/kotlin/zw/co/dcl/jawce/chatbot/service/WebhookConfigService.kt
package zw.co.dcl.jawce.chatbot.service

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import zw.co.dcl.jawce.chatbot.BotConstants
import zw.co.dcl.jawce.chatbot.configs.BotConfigs
import zw.co.dcl.jawce.chatbot.configs.CredentialConfigs
import zw.co.dcl.jawce.engine.model.SessionSettings
import zw.co.dcl.jawce.engine.model.WhatsappSettings
import zw.co.dcl.jawce.engine.model.dto.ChannelOriginConfig
import zw.co.dcl.jawce.engine.model.dto.EngineRequestSettings
import zw.co.dcl.jawce.engine.model.dto.WaEngineConfig
import zw.co.dcl.jawce.engine.service.EntryService
import zw.co.dcl.jawce.session.ISessionManager

@Service
class WebhookConfigService(
    @Qualifier("botTemplates") private val templatesMap: Map<String, Any>,
    @Qualifier("botTriggers") private val triggersMap: Map<String, Any>,
    private val sessionManager: ISessionManager,
    private val botCredentials: CredentialConfigs,
    private val botConfig: BotConfigs
) {

    fun entryInstance(): EntryService {
        val channelOrigin = ChannelOriginConfig(
            false, listOf(), false, null,
            BotConstants.ACCESS_CONTROL_ORIGIN
        )

        return EntryService.getInstance(engineConfig(), channelOrigin)
    }

    private fun engineConfig(): WaEngineConfig {
        val requestSettings = EngineRequestSettings(
            botConfig.botEngineHookBaseUrl,
            botConfig.botEngineHookUrlToken
        )

        return WaEngineConfig(
            sessionManager,
            templatesMap,
            triggersMap,
            requestSettings,
            whatsappChannelSettings(),
            sessionSettings()
        )
    }

    private fun whatsappChannelSettings(): WhatsappSettings {
        var settings = WhatsappSettings()
        settings.hubToken = botCredentials.hubToken
        settings.accessToken = botCredentials.accessToken
        settings.phoneNumberId = botCredentials.phoneNumberId
        settings.apiVersion = BotConstants.API_VERSION
        return settings;
    }

    private fun sessionSettings(): SessionSettings {
        var settings = SessionSettings()
        settings.sessionTTL = botCredentials.sessionTtl ?: 30
        settings.isHandleSessionInactivity = false
        settings.startMenuStageKey = botCredentials.initialStage
        return settings
    }

    fun clearSession(userSessionId: String) {
        sessionManager.clear(userSessionId)
    }
}
